### ISSUE TITLE
We keep introducing bugs to do with time in tests.

### ISSUE THREAD

**Issue by Gnuxie** [matrix-org/mjolnir#327]
There have been previous examples (but i am not going to dig for them yet) this is quite a good one though. https://github.com/matrix-org/mjolnir/pull/325

Let's take a look at what happened. We have a test that looks like (see below). To setup this test we register a new protection that will resolve a promise when we receive a report and this promise resolving will tell us the test has passed, otherwise it will time out. We are only allowed to access the promise resolver from within the [promise executor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise) (well, we can but it would be ugly). So this means we have to register the listener (for when we submit a report) within the executor. Now the problem is that the executor will not run when the promise is created, we have to wait for it to get a spot on the event loop (Disclaimer, i don't understand how this happens, only that control can only be transferred to it while we are awaiting on something else). The point is that we can get to the end of the test (where we await the promise with `await reportPromise` (and i know because this happened), before the promise executor has even run. This means that in the lines before we have already sent a report and Mjolnir has called every registered protection's `handleReport` listener, but because we never registered our protection, our listener was never called and so the `reportPromise` will never resolve.   

```
    it("Mjolnir correctly retrieves a report from synapse", async function() {
        this.timeout(40000);

        const reportPromise = new Promise(async (resolve, reject) => {
            await this.mjolnir.registerProtection(new class implements IProtection {
                name = "jYvufI";
                description = "A test protection";
                settings = { };
                handleEvent = async (mjolnir: Mjolnir, roomId: string, event: any) => { };
                handleReport = (mjolnir: Mjolnir, roomId: string, reporterId: string, event: any, reason?: string) => {
                    if (reason === "x5h1Je") {
                        resolve(null);
                    }
                };
            });
        });
        await this.mjolnir.enableProtection("jYvufI");

        let protectedRoomId = await this.mjolnir.client.createRoom({ invite: [await client.getUserId()] });
        await client.joinRoom(protectedRoomId);
        await this.mjolnir.addProtectedRoom(protectedRoomId);

        const eventId = await client.sendMessage(protectedRoomId, {msgtype: "m.text", body: "uwNd3q"});
        await client.doRequest(
            "POST",
            `/_matrix/client/r0/rooms/${encodeURIComponent(protectedRoomId)}/report/${encodeURIComponent(eventId)}`, "", {
                reason: "x5h1Je"
            }
        );

        await reportPromise;
    });
```
Then once the test has been rewritten so that this doesn't happen 
    
```
        it("Mjolnir correctly retrieves a report from synapse", async function() {
        this.timeout(40000);

        let protectedRoomId = await this.mjolnir.client.createRoom({ invite: [await client.getUserId()] });
        await client.joinRoom(protectedRoomId);
        await this.mjolnir.addProtectedRoom(protectedRoomId);

        const eventId = await client.sendMessage(protectedRoomId, {msgtype: "m.text", body: "uwNd3q"});
        await new Promise(async resolve => {
            await this.mjolnir.registerProtection(new class implements IProtection {
                name = "jYvufI";
                description = "A test protection";
                settings = { };
                handleEvent = async (mjolnir: Mjolnir, roomId: string, event: any) => { };
                handleReport = (mjolnir: Mjolnir, roomId: string, reporterId: string, event: any, reason?: string) => {
                    if (reason === "x5h1Je") {
                        resolve(null);
                    }
                };
            });
            await this.mjolnir.enableProtection("jYvufI");
            await client.doRequest(
                "POST",
                `/_matrix/client/r0/rooms/${encodeURIComponent(protectedRoomId)}/report/${encodeURIComponent(eventId)}`, "", {
                    reason: "x5h1Je"
                }
            );
        });
        // So I kid you not, it seems like we can quit before the webserver for reports sends a respond to the client (via L#26)
        // because the promise above gets resolved before we finish awaiting the report sending request on L#31,
        // then mocha's cleanup code runs (and shuts down the webserver) before the webserver can respond.
        // Wait a minute 😲😲🤯 it's not even supposed to be using the webserver if this is testing report polling.
        // Ok, well apparently that needs a big refactor to change, but if you change the config before running this test,
        // then you can ensure that report polling works. https://github.com/matrix-org/mjolnir/issues/326.
        await new Promise(resolve => setTimeout(resolve, 1000));
    });
```
We get a pretty similar issue where we can kill the webserver that Mjolnir has, before it responds to the request posting a report, which then means that this request is left unanswered until it times out (leading to a long wait or hang if you are running the test) (Note, the report shouldn't be going through the webserver anyways, but that's besides the point). (Note the setTimeout at the end is there so that the webserver has time to respond, but you should be able to remove it and see it happen for yourself)

**Comment by Yoric** [comment]
So if the promise never resolves, we get a timeout error, right? Which at least gives us the opportunity to fix the issue before merging the code, if I understand correctly?

**Comment by Gnuxie** [comment]
Yes it does fail, but it will be intermittent (so it can pass, and in this case it seemed to most of the time) since sometimes the promise executor can run and finish while we await other stuff (e.g. joining protected rooms, sending the message to report)

**Comment by Gnuxie** [comment]
It is not a problem with the test runner at all, it is just the nature of concurrency in JS.

I don't know what you mean, but yes I understand what happens (the only thing I don't understand is how tasks get scheduled on the event loop, but i do not believe it is entirely relevant when we consider that the promise executor itself is async)

The test will terminate with a timeout error (either from mocha or in the case of the webserver, the request sending the report) in all these situations (apart from the one where the test passes).

It's not an issue about waiting for all promises to be resolved rejected, we don't await for the promise that sends a report to the webserver anywhere in the test.

**Comment by Gnuxie** [comment]
Because reportPromise can resolve because most of the time the report does get sent after the listener registered rather than before

**Comment by Gnuxie** [comment]
then in the second version of the test, `reportPromise` does resolve before we finish awaiting the response to the web server when sending the report, which ends the test.

**Comment by Yoric** [comment]
> Because reportPromise can resolve because most of the time the report does get sent after the listener registered rather than before

Ah, so we're failing to detect that there is a bug in the test itself, but it's a bug that cases false negatives rather than false positives, right?

**Comment by Gnuxie** [comment]
No because in the first version of the test, if the report sends before the listener is registered (ie the protection is registered) then we never resolve `reportPromise` because the report is never given to our protection to handle. (Then the test will time out).

**Comment by Yoric** [comment]
Yes, that's what I meant with false negative, but I now realize that it's ambiguous :)
